Cross account IAM deployment

mattmb opened the pull request:
This PR adds several features:
- It adds the ability to authenticate using cross account AWS IAM roles.
- It adds the ability to include custom cloudformation templates.
- It makes the deploy script runnable as a limited user.

The overall goal is that we can create an IAM user in a central AWS account which will get limited permissions by way of an included template. The user will then be able to login using the new cross account authentication. Finally the user will be able to run the salt_utils script as a limited sudoer.

This is we think cleaner than #71 and allows us to move the IAM roles into template deploy with https://github.com/ministryofjustice/template-deploy/pull/23

Commit:
Support AWS role assumption

This adds an extra bit of code to the AWS api connection utility so
that it is possible to assume an IAM role from another account (if your
user has the right to). The ARN id of the role to assume is expected as
an environment variable.

Commit:
Move dict merge to utils.

Moving dictionary deep merge to utils module so that it is easier to
reuse accross classes.

Commit:
Add include functionality

This adds the option to include extra cloudformation json from a static
file. The includes are merged with the template we generate, overwriting
existing keys that clash.

Commit:
Update README/CHANGELOG

Commit:
Append bootstrap-cfn path in salt_utils.py

This adds the path explicitly in the salt_utils module so that a user
with limited access does not have to set their PYTHONPATH.

mattmb commented:
@ashb @stuartornum here's the new approach as we discussed for review.

ashb commented on README.rst:
```diff
@@ -109,3 +109,12 @@ By default the ELBs will have a security group opening them to the world on 80 a
+    includes:
+      - /path/to/cloudformation.json
+
+The tool will then merge this with the generated template *overwriting* any keys in the original template that clash.
```
This isn't quite clear on two counts:

1) What's a key? (Are we talking a matching logical resource id)
2) That we do deep merging (i.e. we're not overwriting the Output section entirely.)

mattmb replied:
Will try to make it clearer. Quite hard to explain deep merging in English though!

ashb commented on bootstrap_cfn/salt_utils.py:
```diff
@@ -1,4 +1,6 @@
 #!/usr/bin/env python
+import sys
+sys.path.append('/usr/local/bootstrap-cfn')
```
Hard coded paths are bad - it makes assumptions about where the code will live in multiple places in the code. I think this `os.path.dirname(__file__)` achieves what we want.

Also can you add a comment saying _why_ we are doing this so we remember in 12 months time.

mattmb replied:
Good shout, actually needs `sys.path.append(os.path.dirname(__file__) + '/..')` I think the folder of the module is always in sys.path.

ashb commented on bootstrap_cfn/config.py:
```diff
@@ -92,6 +76,10 @@ def process(self):
+        if 'includes' in self.data:
+            for inc_path in self.data['includes']:
+                inc = json.load(open(inc_path))
+                utils.dict_merge(template, inc)
```
You aren't doing anything with the return value from `dict_merge` - which means that it's merging in place into template - if that's the case can you remove the return statement from that function and add a simple doc comment.

ashb replied:
Hmmm that wouldn't work actually - as it's recursive and relies on the return value inside itself. Hmmm!

ashb replied:
Just do `template = utilts.dict_merge(template, inc)` I guess.

mattmb replied:
Ugh, I don't want to think to hard about why this seems to work because thinking about recursive functions always gives me a headache! I think there might be an edge case where it doesn't work if you don't assign the return value though so I will change it.

ashb commented:
The few minor comments above not withstanding this PR actually does three completely separate things
1. It adds the ability to include static snippets of Cloudformation JSON
2. It lets you use the sts AssumeRole to target another account
3. It fixes the PYTHONPATH issue in the deploy script.

These happen to be related since you want them all, but even your commits show that it's three features (the first three commits are point 1, next is point 2 and last is point 3). Sorry to be a pain but can you split into three different PRs please?

ashb commented:
(Or at least come up with a better PR title encompasses all the features)

mattmb commented:
Yeah, wasn't sure whether to split or not. I feel like it's kind of one story so nice to have in one place but they are quite distinct features. I think there's some good history in this PR, so I'd rather leave it be if you're happy for me to rename it and tidy up the description :-)

Commit:
Fix include merging.

This commit fixes a couple of mistakes in the merging code for included
cloudformation templates. It also makes the documentation clearer on how
the merge works.

ashb commented:
Title is much better now :+1:.

It looks like you need to rebase this onto master at least - possibly squash a few of the commits too in the process (but not all of them)

Also can you document AWS_ROLE_ARN_ID (and possibly how you would get it) and add a changelog entry for this feature.

mattmb commented:
@ashb rebased and tidy up :smile:
